Adding a rough draft of the "minimum viable product" for the .NET Libraries APIs to support generic math (#205)
dotnet/designs: accepted/2021/statics-in-interfaces/README.md

C# is looking at enabling static abstract members in interfaces (https://github.com/dotnet/csharplang/issues/4436). From the libraries perspective, this is an opportunity to enable "generic math" which could define a new baseline for how developers write algorithms in .NET.

## Scenarios

> ### Allow easier maintainence for generic code

Reviewer: ```suggestion
### Allow easier maintenance for generic code
```

    }
    else if (typeof(T) == typeof(ushort))
    {
        return (T)(object)(ushort)((ushort)(object)left + (ushort)(object)right);
    }
    else if (typeof(T) == typeof(short))
    {
        return (T)(object)(short)((short)(object)left + (short)(object)right);
    }
    else if (typeof(T) == typeof(uint))
    {
        return (T)(object)(uint)((uint)(object)left + (uint)(object)right);
    }
    else if (typeof(T) == typeof(int))
    {
        return (T)(object)(int)((int)(object)left + (int)(object)right);
    }
    else if (typeof(T) == typeof(ulong))
    {
        return (T)(object)(ulong)((ulong)(object)left + (ulong)(object)right);
    }
    else if (typeof(T) == typeof(long))
    {
        return (T)(object)(long)((long)(object)left + (long)(object)right);
    }
    else if (typeof(T) == typeof(float))
    {
        return (T)(object)(float)((float)(object)left + (float)(object)right);
    }
    else if (typeof(T) == typeof(double))
    {
        return (T)(object)(double)((double)(object)left + (double)(object)right);
    }
    else
    {
        throw new NotSupportedException(SR.Arg_TypeNotSupported);
    }
}
```

> Given the operation is the same for each type, only differing on the actual type in question, Tanner would expect this to be easier to author and maintain. Additionlly, each new type that needs to be supported adds additional branches and complexity to the code. For example, in .NET 6 support for `nint` and `nuint` support is desired.

Reviewer: ```suggestion
Given the operation is the same for each type, only differing on the actual type in question, Tanner would expect this to be easier to author and maintain than this. Additionally, each new type that needs to be supported adds additional branches and complexity to the code. For example, in .NET 6 support for `nint` and `nuint` support is desired.
```

```csharp
public interface IAddable<TSelf>
    where TSelf : IAddable<TSelf>
{
    static abstract TSelf operator +(TSelf left, TSelf right);
}

public interface ISubtractable<TSelf>
    where TSelf : ISubtractable<TSelf>
{
    static abstract TSelf operator -(TSelf left, TSelf right);
}

public interface IMultipliable<TSelf>
    where TSelf : IMultipliable<TSelf>
{
    static abstract TSelf operator -(TSelf left, TSelf right);
}

public interface IDivisable<TSelf>
    where TSelf : IDivisable<TSelf>
{
    static abstract TSelf operator /(TSelf left, TSelf right);
}

public interface INumber<TSelf>
    : IAddable<TSelf>
    : ISubtractable<TSelf>
    : IMultipliable<TSelf>
    : IDivisable<TSelf>
    where TSelf : INumber<TSelf>
{
}

// New methods on System.Linq.Enumerable for `Sum<T>` and `Average<T>`, `where T : INumber<T>`
```

Jeff can now implement their method such as:
```csharp
public static double GetStandardDeviation<T>(this IEnumerable<T> values)
    where T : INumber<T>
{
    double standardDeviation = 0;
    T[] enumerable = values as T[] ?? values.ToArray();
    int count = enumerable.Count();
    if (count > 1)
    {
        double avg = enumerable.Average();

>         double sum = enumerable.Sum(d => (d - avg) * (d - avg));

Author: I forgot to put in the conversion to double here. It would be:
```suggestion
        double sum = enumerable.Sum(d => ((double)d - avg) * ((double)d - avg));
```

(or `double sum = enumerable.Sum(d => (double.CreateFrom(d) - avg) * (double.CreateFrom(d) - avg));`)

Some of the code below assumes that https://github.com/dotnet/csharplang/issues/4665 is going to be accespted into the language. If it is not accepted or if the approved syntax differs, the declarations will need to be updated.

```csharp
namespace System
{
    public interface IParseable<TSelf>
        where TSelf : IParseable<TSelf>
    {
        // These do not take NumberStyles as not all parseable types are numbers (such as Guid)
        // Instead, INumber<TSelf> exposes additional parse methods that do take NumberStyles

        static abstract TSelf Parse(string s, IFormatProvider? provider);

        static abstract bool TryParse(string s, IFormatProvider? provider, out TSelf result);
    }

    public interface ISpanParseable<TSelf> : IParseable<TSelf>
        where TSelf : ISpanParseable<TSelf>
    {

>         // This inherits from IParseable, much as ISpanFormattable is planned to inherit from IFormattable

Author: No, we don't strictly need both and could have just `IParseable` and have it expose the `Span` overloads.
* This might negatively impact other languages and has the general downsides of exposing just span overloads, however

I figured that is a question for API review, however. I also believe DIM support in v1 is still up in the air.

>         // This inherits from IParseable, much as ISpanFormattable is planned to inherit from IFormattable

Reviewer 2: > This might negatively impact other languages and has the general downsides of exposing just span overloads, however

From my perspective, so be it :smile:.  We're introducing a brand new feature here in static interfaces that no one currently supports, so to even hit an issue with spans you'd first need to update your language to support this large new capability.  At that point, supports spans, too, or at least support passing the result of `string.AsSpan()` instead of `string` as an argument.

        static abstract TSelf Parse(ReadOnlySpan<char> s, IFormatProvider? provider);

        static abstract bool TryParse(ReadOnlySpan<char> s, IFormatProvider? provider, out TSelf result);
    }

    public interface IEquatableOperators<TSelf, TOther> : IEquatable<TOther>
        where TSelf : IEquatableOperators<TSelf, TOther>
    {

>         // Given this takes two generic parameters, we could simply call it IEquatable<TSelf, TOther>

Author: API review went with `I*Operators` as the name for all operator interfaces. So this is named `IEqualityOperators` now.

        // This inherits from IEquatableOperators<TSelf, TOther> as even though IComparable<T> does not
        // inherit from IEquatable<T>, the <= and >= operators as well as CompareTo itself imply equality

        static abstract bool operator <(TSelf lhs, TOther rhs);

        static abstract bool operator <=(TSelf lhs, TOther rhs);

        static abstract bool operator >(TSelf lhs, TOther rhs);

        static abstract bool operator >=(TSelf lhs, TOther rhs);
    }

    public interface IAddable<TSelf, TOther, TResult>
        where TSelf : IAddable<TSelf, TOther, TResult>
    {
        // An additive identity is not exposed here as it may be problematic for cases like
        // DateTime that implements both IAddable<DateTime, DateTime> and IAddable<DateTime, TimeSpan>
        // This would require overloading by return type and may be problematic for some scenarios where TOther might not have an additive identity

        // We can't assume TOther + TSelf is valid as not everything is commutative

        // We can't expose TSelf - TOther as some types only support one of the operations

        static abstract TResult operator +(TSelf lhs, TOther rhs);

>         static abstract TResult checked operator +(TSelf lhs, TOther rhs);

Author: Yes, there are a number of cases where it would be convenient to use DIMs.

However, I believe DIM support in v1 is still up in the air. We've listed it as a "requirement" for future versioning (so v2+).

        where TSelf : IMultipliable<TSelf, TOther, TResult>
    {
        // A multiplicative identity is not exposed here for the same reasons as IAddable<TSelf, TOther>

        // We can't assume TOther * TSelf is valid as not everything is commutative

        // We can't expose TSelf / TOther as some types, such as matrices, don't support dividing with respect to TOther

        // We can't inherit from IAddable<TSelf, TOther> as some types, such as matrices, support Matrix * scalar but not Matrix + scalar

        static abstract TResult operator *(TSelf lhs, TOther rhs);

        static abstract TResult checked operator *(TSelf lhs, TOther rhs);
    }

    public interface IDivisable<TSelf, TOther, TResult>
        where TSelf : IDivisable<TSelf, TOther, TResult>
    {
        static abstract TResult operator /(TSelf lhs, TOther rhs);

        static abstract TResult checked operator /(TSelf lhs, TOther rhs);
    }

    public interface IRemainder<TSelf, TOther, TResult>
        where TSelf : IRemainder<TSelf, TOther, TResult>
    {
        // The ECMA name is op_Modulus and so the behavior isn't actually "remainder" with regards to negative values

>         // Likewise the name IRemainder doesn't fit with the other names, so a better one is needed

Reviewer: IRemainderable? I assume it would be implemented by `%` for the built in types.

It may be called op_Modulus in ECMA but as far as I can see it's performing the remainder, hence the name of the C# `%` operator. 
https://stackoverflow.com/questions/13683563/whats-the-difference-between-mod-and-remainder

Reviewer: Are there types that are divisible but not remainderable?

Reviewer 3: Ha -- I asked the inverse :-)
https://github.com/dotnet/designs/pull/205#discussion_r613744151

Reviewer: Ah yes - that is already addressed then 🙂

        static abstract TResult operator %(TSelf lhs, TOther rhs);

        static abstract TResult checked operator %(TSelf lhs, TOther rhs);

>     }

Author: It's further down: https://github.com/dotnet/designs/pull/205/files#diff-4dfd0bc83bc94153a6e400acf779e0536171cc8ff6e5e0016eeab62985da365eR600

    public interface INegatable<TSelf, TResult>
        where TSelf : INegatable<TSelf, TResult>
    {

>         // Should unary plus be on its own type?

Reviewer 4: If not, this seems like a good case for supporting default implementations since (I think it's safe to say) 99% of implementations of unary `+` will just return `value` unmodified.

        static abstract TResult operator +(TSelf value);

>         static abstract TResult checked operator +(TSelf value);

Author: > I'm wondering why this operator+ needs to exist.

For parity with `-`. There are libraries that use `+` for correctness, alignment, and readability.

Unfortunately, its a bit late for the language to just implicitly provide it :(

>         static abstract TResult checked operator +(TSelf value);

Reviewer 2: > There are libraries that use + for correctness, alignment, and readability.

I understand "readability".  Can you elaborate on how `+value` vs `value` impacts correctness?  And I'm not sure what you mean here by "alignment"?

>         static abstract TResult checked operator +(TSelf value);

Author: It allows alignment of values without fighting the formatter:
```csharp
const float x = 1.234;
const float y = -1.234;
const float z = +1.234;
```

As for correctness, once you start getting into heavy math code, you'll often end up writing series where every step involves a constant with an alternating positive/negative sign.

For example, `Sin` and `Cos` are:
![image](https://user-images.githubusercontent.com/10487869/115889205-961b5780-a408-11eb-8acc-2d133ccf4447.png)

In which case you will likely have a lookup table, where the entries are different constants that have alternating `+` vs `-` and where explicitly annotating the `+` helps with show "this constant was meant to be positive"

>         static abstract TResult checked operator +(TSelf value);

Reviewer 2: Ok, so not correctness in that it's actually impacting functionality.  You just meant having + allows someone to write code that looks and reads better and thus may help them to avoid bugs.

I'm not convinced that's worth requiring every type to have another two interface methods that are always going to just return the same value.  But open for discussion I guess.

        static abstract TResult operator -(TSelf value);

        static abstract TResult checked operator -(TSelf value);
    }

    public interface IFixedWidth<TSelf>
        where TSelf : IFixedWidth<TSelf>
    {
        // MinValue and MaxValue are more general purpose than just "numbers", so they live on this common base type

>         // Fixed-Width isn't the best name however, so we should think of alternatives

Reviewer: I think Bounded is the [mathematical term](https://en.wikipedia.org/wiki/Bounded_set)

>         // Fixed-Width isn't the best name however, so we should think of alternatives

Author: We went with `IMinMaxValue` based on the API review.

        static abstract TSelf MinValue { get; }

        static abstract TSelf MaxValue { get; }
    }
}
```

### Numeric Interfaces

The numeric interfaces build upon the base interfaces by defining the core abstraction that most users are expected to interact with.

```csharp
namespace System
{
    public interface INumber<TSelf>
        : IAddable<TSelf, TSelf>,
          IComparableOperators<TSelf, TSelf>,   // implies IEquatableOperators<TSelf, TSelf>
          IDecrementable<TSelf>,
          IDivisable<TSelf, TSelf>,
          IIncrementable<TSelf>,
          IMultipliable<TSelf, TSelf>,
          IRemainder<TSelf, TSelf>,
          ISpanFormattable,                     // implies IFormattable
          ISpanParseable<TSelf>,                // implies IParseable<TSelf>
          ISubtractable<TSelf, TSelf>
        where TSelf : INumber<TSelf>
    {
        // For the Create methods, there is some concern over users implementing them. It is not necessarily trivial to take an arbitrary TOther
        // and convert it into an arbitrary TSelf. If you know the type of TOther, you may be able to optimize in a few ways. Otherwise, you

>         // may have to fail and throw in the worst case.

Author: I think it entirely depends on the types you are wanting to support and consume.

I'd expect it can succeed most of the time, but may not be fast for uncommon types. You would only need to fail if a user provided some type for which there is no way to get the necessary value to use in the conversion.

>         // may have to fail and throw in the worst case.

Author: I think in either scenario we will need to support looking for conversions on both `TSelf` and `TOther`

We could also define some `IConvertible` interface to help with this, but its something I'm still experimenting with while trying to write scenarios.

>         // may have to fail and throw in the worst case.

Reviewer 5: I have not read everything here but having explicit support for casts/conversion sounds like a better way for this, in many algorithms for image processing you want to move from say `byte` to `short`/`int` and having interfaces for that seems like a good idea.

        static abstract TSelf Create<TOther>(TOther value)
            where TOther : INumber<TOther>;

        static abstract TSelf? CreateExactly<TOther>(TOther value)
            where TOther : INumber<TOther>;

        static abstract TSelf CreateSaturating<TOther>(TOther value)
            where TOther : INumber<TOther>;

        static abstract TSelf CreateTruncating<TOther>(TOther value)
            where TOther : INumber<TOther>;

        // There is an open question on whether properties like IsSigned, IsBinary, IsFixedWidth, Base/Radix, and others are beneficial
        // We could expose them for trivial checks or users could be required to check for the corresponding correct interfaces

        static abstract TSelf One { get; }

>         static abstract TSelf Zero { get; }

Reviewer 6: Would it be possible to move these up to some of the parent interfaces? For example, move `Zero` to `IAddable<T>` and move `One` to `IMultipliable<T>`?

There are mathematical types (for example Matrices or Elliptic Curves) where they have identity elements for the operations they support, but they don't have support for additional operations (ex. Matrices doesn't support Division but has a Zero and One. Elliptic Curves have a Zero and addition but don't have multiplication or a One element)

Author: Not easily. You have to consider whether it should be `TSelf One/Zero` or `TOther One/Zero` or `TResult One/Zero`. Likewise, not everything has a "clear" concept of `one` vs `zero`. It also gets problematic when considering other types like `ISubtractable` or `IDivisible` as well as  when considering types that implement multiple `IAddable` or `IMultipliable` interfaces (differing on `TOther` at least), The latter of which requires you to consider if all concepts of `One` are equivalent.

Instead, my current thinking is that there will need to be a different interface for "sets of numbers" (which includes vectors, matrices, tensors, etc). This might simply be `IVector<T>`, `IMatrix<T>`, and `ITensor<T>`, but its something I'm actively trying to get a design for.

Reviewer 6: Yes, the identity element only makes sense for closed operations where `TSelf`, `TOther`, and `TResult`, are all the same type. So they'd require a one-parameter version of `IAddable<T>` and `IMulitpliable<T>` to make sense.

Even if there were other interfaces like `IAddableIdentity<T>` and `IMultipliableIdentity<T>`, I'd be happy with that. I want to be able to write generic math that works at a group/ring level and I can't do that if the identity element property is specific to numbers, or vectors, or matrices, or curves, etc.

Author: I don't think we are going to be exposing concepts like `group`/`ring` or `monoid`.

We must expose these concepts via interfaces, not traits, and there are far too many existing .NET types and concepts that do not fit the mathematical model you find in many functional languages, like Haskell.

There is also a limit on the number of interfaces we can add here. The number of interfaces currently modeled here is already "a lot" and is expected to have a measurable impact to application startup.

Limitations in the runtime also come into play here:
* We can't overload based on generic constraint
* You can only use these via constrained generics (because these must be a `constrained call`)
* Changing your generic constraint is a breaking change
* There is no concept of associated types
* etc

All of these add up to defining things more like `IComparable` or `IEquatable`, which expose a general contract for a given operation but don't provide more than a few guarantees and so `IAddable` will provide a base interface for building higher abstractions, like `INumber` or `IVector` and most code will need to take types like `INumber` or `IVector` and some very specialized methods can take a base interface like `IAddable`.

Some concepts like `Commutativity`, `Associativity`, or `Distributivity` might be possible to light up for tools via attributes. It will likely not be a first class citizen due to JIT time constraints and cases like `float` where most of the rules don't exist in IEEE 754 compliant code.

Author: We will certainly discuss this in API review and I will be raising what might be left on the table if we don't expose these concepts in v1.

But as of right now, these concepts are not trivial to model for C#-like .NET code, which will be the primary user of these

Reviewer 6: One example to definitely think about is that you can't write a `Sum` function for `IEnumerable<IAddable<T, U, V>>` without either having an identity element on the interface or requiring users to pass one in.

Author: After a lot of thought, I added `IAdditiveIdentity` and `IMultiplicativeIdentity` to fill this gap as I believe its the simplest way to cover it and fit in with the existing model.

        // Abs mirrors Math.Abs and returns the same type. This can fail for MinValue of signed integer types
        // Swift has an associated type that can be used here, which would require an additional type parameter in .NET
        // However that would hinder the reusability of these interfaces in constraints

        static abstract TSelf Abs(TSelf value);

        static abstract TSelf Clamp(TSelf value, TSelf min, TSelf max);

>         static abstract (TSelf Quotient, TSelf Remainder) DivRem(TSelf left, TSelf right);

Author: This depends on if the language/runtime ship support for DIMs in v1 or v2 and if there is a way for users to even invoke or forward to the DIM.

        // them in SIMD contexts today, as do most languages with SIMD support, so there is not great concern here.

        static abstract TSelf operator &(TSelf lhs, TSelf rhs);

        static abstract TSelf operator |(TSelf lhs, TSelf rhs);

        static abstract TSelf operator ^(TSelf lhs, TSelf rhs);

        static abstract TSelf operator ~(TSelf value);

        // The shifting operators should probably be defined on some common base interface, like IShiftable
        // There is an open question on if we need language support for "arithmetic" vs "logical" right shift"
        // ECMA 335 already defines `op_SignedRightShift` and `op_UnsignedRightShift`

        // Likewise, forcing int here is inconvenient from the perspective of generic code where the amount will
        // be "in range" (e.g. 0-63 for int64) but where it may be the result of other logic and not be an int32

        static abstract TSelf operator <<(TSelf value, int shiftAmount);

        static abstract TSelf operator >>(TSelf value, int shiftAmount);

        static abstract bool IsPow2();

>         static abstract TSelf Log2();

Author: `Math` and `MathF` have always been unusual as almost all non-primitive types expose their "operations" on the type themselves. This is true for `DateTime`, `TimeSpan`, `Complex`, most of `Decimal`, `BigInteger`, `Vector<T>`, `Vector2/3/4`, `Plane`, `Quaternion`, etc
`Math` and `MathF` also have versioning issues (which is why `MathF` exists in the first place) and concerns are always raised about user confusion between the `integral` and `floating-point` method differences.

So, I'd be fine if we move towards just exposing types on `IFloatingPoint` moving forward, but this is likely a much broader discussion in API review.

        // We have a "well-defined" behavior for the next few methods in BigInteger. There are some cases where we would likely benefit
        // returning nint/nuint to better support big data and its possible we would want to break compat here for a better experience

        long GetBitLength();

        int GetByteCount(bool isUnsigned);

        byte[] ToByteArray()

>         byte[] ToByteArray(bool isUnsigned = false, bool isBigEndian = false);

Author: This was copied from `BigInteger` today as the "initial rough draft". I think it needs a few tweaks and scenarios showing its use in practice.

        // If we want decimal to implement it, we would need to move several methods down and define
        // some IeeeFloatingPoint interface instead. This wasn't done for simplicity and because decimal
        // is not an industry standard type.

        // TODO: The Exponent and Signifcand need to be exposed here
        // They should return INumber<TSelf> here and IBinaryNumber<TSelf> on IBinaryFloatingPoint<TSelf>

        // TODO: We may want to expose constants related to the subnormal and finite boundaries

        // The following constants are defined by System.Math and System.MathF today

        static abstract TSelf E { get; }

        static abstract TSelf PI { get; }

        static abstract TSelf Tau { get; }

        // The following methods are defined by System.Math and System.MathF today
        // Exposing them on the interfaces means they will become available as statics on the primitive types
        // This, in a way, obsoletes Math/MathF and brings float/double inline with how non-primitive types support similar functionality
        // API review will need to determine if we'll want to continue adding APIs to Math/MathF in the future

>         static abstract TSelf Acos(TSelf x);

Reviewer 2: Why are many of these static rather than instance?  i.e. I'm not sure why this isn't just `TSelf Acos();`?

Author: Because these kinds of operations have traditionally been `static` in .NET.

Also, there is a perf consideration here because `this` is `byref` (or `inref` for `readonly struct`) which can hurt inlining and other optimizations done by the JIT.

        : IBinaryNumber<TSelf>,
          IFloatingPoint<TSelf>
        where TSelf : IBinaryFloatingPoint<TSelf>
    {
    }

    public interface IDecimalFloatingPoint<TSelf>
        : IFloatingPoint<TSelf>
        where TSelf : IDecimalFloatingPoint<TSelf>
    {
        // This interface is defined for convenience of viewing the IEEE requirements
        // it would not actually be defined until the .NET libraries requires it

        // The majority of the IEEE required operations are listed below
        // This doesn't include the recommended operations such as sin, cos, acos, etc

        // 5.3.2
        //  TSelf Quantize(TSelf x, TSelf y);
        //  TSelf Quantum(TSelf x);

        // 5.5.2
        //  TDecEnc EncodeDecimal(TSelf x);
        //  TSelf DecodeDecimal(TDecEnc x);
        //  TBinEnc EncodeBinary(TSelf x);
        //  TSelf DecodeBinary(TBinEnc x);

        // 5.7.3
        //  bool SameQuantum(TSelf x, TSelf y);
    }
}
```

### Interface Implementors

Various types in the System namespace would implement the above interfaces as follows.

```csharp
namespace System
{
    public struct Byte
        : IBinaryInteger<byte>,

>           IConvertible,

Reviewer 2: Might be worth denoting somehow that some of these interfaces are already implemented by these types.

          IComparableOperators<TimeSpan, TimeSpan>,
          IDivisable<TimeSpan, double, TimeSpan>,
          IDivisable<TimeSpan, TimeSpan, double>,
          IFixedWidth<TimeSpan>,
          IMultipliable<TimeSpan, double, TimeSpan>,
          INegatable<TimeSpan, TResult>,
          ISpanFormattable,
          ISubtractable<TimeSpan, TimeSpan, TimeSpan>
    {
    }

    public struct UInt16
        : IBinaryInteger<ushort>,
          IConvertible,
          IFixedWidth<ushort>,
          IUnsignedNumber<ushort>
    {
    }

    public struct UInt32
        : IBinaryInteger<uint>,
          IConvertible,
          IFixedWidth<uint>,
          IUnsignedNumber<uint>
    {
    }

    public struct UInt64
        : IBinaryInteger<ulong>,
          IConvertible,
          IFixedWidth<ulong>,
          IUnsignedNumber<ulong>
    {
    }

    public struct UIntPtr
        : IBinaryInteger<nuint>,
          IFixedWidth<nuint>,
          ISerializable,
          IUnsignedNumber<nuint>
    {
    }
}
```

### Pending Concepts

There are several types which may benefit from some interface support. These include, but aren't limited to:
* System.Array

> * System.DateOnly

Reviewer 2: You already included this one above.  Same with TimeOnly.

Author: Forgot to remove them from this list when I added them above. Will fix.

* System.Enum
* System.Index
* System.Range
* System.TimeOnly
* System.Tuple
* System.ValueTuple

> * System.Numerics.BigInteger

Reviewer 2: I'd have expected this one to be a given.

Author: Several of them are "given", but I haven't done any scenarios for them yet and so am just tracking them below until I can finish writing the code locally.

* System.Numerics.Complex
* System.Numerics.Matrix3x2

> * System.Numerics.Matrix4x4

Reviewer 2: There are lots of comments throughout the interfaces about how various decisions are being made in order to support types like matrices, so I'd hope we either implement the interfaces on matrices or don't constrain ourselves by things we aren't ourselves willing to do :)

Author: This list is more a `todo` than a "maybe we should think about these".

Even if we decided to not support our own types, for whatever reason, ensuring the interfaces work for 3rd party types is an important scenario.
(I do plan on defining interfaces and including them in the design doc here, but since I only got compiler bits a couple days ago, it was relatively slow going to "validate" the designs 😄)

After thinking over the possible implementations, Jeff realizes that the only reliable way to implement this in .NET 5 is to define an interface with instance methods that users must then implement on their own types or to hardcode a list of types in their own library. They do not believe this to be as reusable or maintainable as hoped.

Jeff learns that a future version of .NET will offer a new feature called static abstracts in interfaces. Such a feature will have the .NET Libraries expose a set of interfaces that allow Jeff to implement the desired method without requiring users types to take a dependency on types they declared.

The .NET Libraries are exposing interfaces such as:
```csharp
public interface IAdditionOperators<TSelf>
    where TSelf : IAdditionOperators<TSelf>
{
    static abstract TSelf operator +(TSelf left, TSelf right);
}

public interface ISubtractionOperators<TSelf>
    where TSelf : ISubtractionOperators<TSelf>
{
    static abstract TSelf operator -(TSelf left, TSelf right);
}

public interface IMultiplyOperators<TSelf>
    where TSelf : IMultiplyOperators<TSelf>
{

>     static abstract TSelf operator -(TSelf left, TSelf right);

Reviewer 7: ```suggestion
    static abstract TSelf operator *(TSelf left, TSelf right);
```

    {
        // The ECMA name is op_Modulus and so the behavior isn't actually "remainder" with regards to negative values

        // Likewise the name IModulusOperators doesn't fit with the other names, so a better one is needed

        static abstract TResult operator %(TSelf left, TOther right);

        static abstract TResult checked operator %(TSelf left, TOther right);
    }

    public interface IUnaryNegationOperators<TSelf, TResult>
        where TSelf : IUnaryNegationOperators<TSelf, TResult>
    {
        // Should unary plus be on its own type?

        static abstract TResult operator +(TSelf value);

        static abstract TResult checked operator +(TSelf value);

        static abstract TResult operator -(TSelf value);

        static abstract TResult checked operator -(TSelf value);
    }

    public interface IBitwiseOperators<TSelf, TOther, TResult>
    {
        static abstract TResult operator &(TSelf left, TOther right);

        static abstract TResult operator |(TSelf left, TOther right);

        static abstract TResult operator ^(TSelf left, TOther right);

        static abstract TResult operator ~(TSelf value);
    }

>     public interface IShiftOperators<TSelf, TOther, TResult>

Author: ```suggestion
    public interface IShiftOperators<TSelf, TOther, TResult>
        where TSelf : IShiftOperators<TSelf, TOther, TResult>
```

        static abstract TSelf CreateTruncating<TOther>(TOther value)
            where TOther : INumber<TOther>;

        // There is an open question on whether properties like IsSigned, IsBinary, IsFixedWidth, Base/Radix, and others are beneficial
        // We could expose them for trivial checks or users could be required to check for the corresponding correct interfaces

        // Alias for MultiplicativeIdentity
        static abstract TSelf One { get; }

        // Alias for AdditiveIdentity
        static abstract TSelf Zero { get; }

        // Abs mirrors Math.Abs and returns the same type. This can fail for MinValue of signed integer types
        // Swift has an associated type that can be used here, which would require an additional type parameter in .NET
        // However that would hinder the reusability of these interfaces in constraints

        static abstract TSelf Abs(TSelf value);

        static abstract TSelf Clamp(TSelf value, TSelf min, TSelf max);

        static abstract (TSelf Quotient, TSelf Remainder) DivRem(TSelf left, TSelf right);

        static abstract TSelf Max(TSelf x, TSelf y);

        static abstract TSelf Min(TSelf x, TSelf y);

> 

Author: ```suggestion

        static abstract TSelf Parse(string s, NumberStyles style, IFormatProvider? provider);

        static abstract TSelf Parse(ReadOnlySpan<char> s, NumberStyles style, IFormatProvider? provider);

        static abstract bool TryParse([NotNullWhen(true)] string? s, NumberStyles style, IFormatProvider? provider, out TSelf result);

        static abstract bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatProvider? provider, out TSelf result);

```

    {
    }
}

namespace System.Runtime.Intrinsics
{
    public struct Vector64<T> : IVector<Vector64<T>, T>
    {
    }

    public struct Vector128<T> : IVector<Vector128<T>, T>
    {
    }

    public struct Vector256<T> : IVector<Vector256<T>, T>
    {
    }
}
```

### Pending Concepts

There are several types which may benefit from some interface support. These include, but aren't limited to:
* System.Array
* System.Enum
* System.Index
* System.Range
* System.Tuple
* System.ValueTuple
* System.Numerics.BigInteger
* System.Numerics.Complex
* System.Numerics.Matrix3x2
* System.Numerics.Matrix4x4
* System.Numerics.Plane
* System.Numerics.Quaternion

Likewise, there are many comments within the defined interfaces above that call out key points that need additional modeling/consideration. Several of these concepts will be determined via user research and outreach while others will be determined via API review and feedback from other area experts. Others may be determined by language or runtime limitations in what is feasible for them to support.

Do we want a way to denote "primitives"? If we have it, should this be language primitives, runtime primitives, or ABI primitives?

> Do we want an explicit `IConvertible<TSelf, TOther>` (or `IConvertibleFrom`/`IConvertibleTo`)?

Reviewer 5: Guess this comment relates to the comment I gave above. I would think yes.

Author: (https://github.com/dotnet/designs/pull/205#discussion_r638217591)
> I have not read everything here but having explicit support for casts/conversion sounds like a better way for this, in many algorithms for image processing you want to move from say byte to short/int and having interfaces for that seems like a good idea.

One issue is that that involves `n` additional interfaces on the primitive types. Each interface implemented by the primitive types will negatively impact startup.
Likewise, there is no way to define say `explicit operator TSelf<TOther>(TOther value) where TOther : INumber<TOther>` and so you can't limit it to say 1 additional interface.

`T.Create(value)` is certainly not as nice as `(T)value`, however it works today and otherwise has all the same "problems" with having to deal with `X.ConvertFrom(y)` vs `y.ConvertTo<X>()`, but with the "added" benefit that you can cover more types of creation (such as exact vs truncating vs saturating, etc).

Reviewer: @davidwrighton @jkotas I know optimizations have been done for the interface lookup perf trap. Do we believe there is more we can do there? It seems a shame to compromise design for something that we could potentially optimize during the lifetime of the types.

Author: Even if the number of interfaces wasn't a concern, there is still the issue of not being able to define `explicit operator TSelf<TOther>(TOther value) where TOther : INumber<TOther>`. This might just be a language request however.

Reviewer 8: > @davidwrighton @jkotas I know optimizations have been done for the interface lookup perf trap. 

I do not think that this applies here. The generic math feature is using the interfaces as constrains that are expected to be resolved at JIT time. I do not expect there are going to have dynamic casts for generic math in steady state. (Also, we have added cache for the dynamic interface casts that is about as much as we can do for those.)

The way that the generic math feature is designed adds non-trivial static footprint that number of scenarios are going to pay for, even if they do not use it. I am not sure what to additional cost of ConvertFrom/ConvertTo would be, but I suspect it would not be that much given the total bill.

Reviewer 9: This isn't interface lookup, but its more about interface implementation, and the vast number of interfaces involved here. Its just very costly to implement with our current model of how interfaces are loaded at type load time.

There are 3 major costs.
1. Allocation of unique types. Each interface loaded one of these is several hundred bytes.
2. Association of interface types with concrete types like System.Int32. 
3. Sheer volume of methods in System.Private.CoreLib. If we build some sort of IConvertible<TSelf, TOther> interface and make all the simple numeric (signed and unsigned integer types + float+ double) types implement the full cross product of them, we're talking 10^2 different interfaces, and the same number of additional methods. Each method is something like 100 bytes in the file, so we're looking at a somewhat unfortunate combinatorial expansion problem.

Reviewer 10: One concern I have with using the `Create<TOther>` pattern over dedicated conversion interfaces is that it is not extensible, and misuse is a runtime failure since [the actual support for `TOther` depends on the implementation](https://github.com/dotnet/runtime/blob/c0662e8129beaf93b8050d39a863cc6d16a0308c/src/libraries/System.Private.CoreLib/src/System/Int32.cs#L488).

For example, consider the following function:

```csharp
static TTo Convert<TFrom, TTo>(TFrom from)
    where TTo : INumber<TTo>
    where TFrom : INumber<TFrom>
    => TTo.Create(from);
```

This works fine for types explicitly supported by `TTo`, IE: `Convert<short, int>(x)`

But if I define my own `INumber` implementation, say an `Int128` type and try `Convert<Int128, int>(x)` it will compile just fine but will fail at runtime with `NotSupportedException`.

Maybe `Create<TOther>` should be restricted to the primitive numeric types?

### Base Interfaces

The base interfaces each describe one or more core operations that are used to build up the core interfaces listed later. In many cases, such as for addition, subtraction, multiplication, and division, there is only a single operator per interface. This split is important in order for them to be reusable against downstream types, like `DateTime` where addition does not imply subtraction or `Matrix` where multiplication does not imply division. These interfaces are not expected to be used by many users and instead primarily by developers looking at defining their own abstractions, as such we are not concerned with adding simplifying interfaces like `IAddable<TSelf> : IAddable<TSelf, TSelf>`. Likewise, there is not a concern with exposing `IAddable<TSelf, TOther, TResult>` and instead we will require `TSelf` and `TResult` be the same. This requirement greatly simplifies implementation and consumption of the types.

> ```

Reviewer 3: ```suggestion
```csharp
```

        // use incrementing for numbers, they are sometimes used by iterators for example

        // We can't expose TSelf-- as some types only support one of the operations

        static abstract TSelf operator ++(TSelf value);
    }

    public interface IDecrementable<TSelf>
        where TSelf : IDecrementable<TSelf>
    {
        static abstract TSelf operator --(TSelf value);
    }

    public interface IMultipliable<TSelf, TOther>
        where TSelf : IMultipliable<TSelf, TOther>
    {
        // An multiplicative identity is not exposed here for the same reasons as IAddable<TSelf, TOther>

        // We can't assume TOther * TSelf is valid as not everything is commutative

        // We can't expose TSelf / TOther as some types, such as matrices, don't support dividing with respect to TOther

        // We can't inherit from IAddable<TSelf, TOther> as some types, such as matrices, support Matrix * scalar but not Matrix + scalar

        static abstract TSelf operator *(TSelf lhs, TOther rhs);
    }

    public interface IDivisable<TSelf, TOther>
        where TSelf : IDivisable<TSelf, TOther>
    {
        static abstract TSelf operator /(TSelf lhs, TOther rhs);
    }

>     public interface IRemainder<TSelf, TOther>

Reviewer 3: Are there cases where something is "remaindable" without being divisible?

Author: Not that I've encountered nor that I can think of off the top of my head. So it might be possible this can inherit from `IDivisable` (C# does not require them to be paired, however, so it is possible).

Reviewer 3: I should have also asked if the inverse is possible? Can something be divisible without supporting the concept of a remainder? I'm wondering if we could just collapse into `IDivisible` with both operators.

Author: Yes, there are cases like Matrix or even TimeSpan that only support division today. We could potentially extend them to also support remainder, but it seems like a restriction that might not apply in practice.

Reviewer 11: Technically numbers that support remainders are known as Euclidean domains. Calling it IEuclidean is probably a stretch but how about IModularizable?

        // There is an open question on whether properties like IsSigned, IsBinary, IsFixedWidth, Base/Radix, and others are beneficial
        // We could expose them for trivial checks or users could be required to check for the corresponding correct interfaces

        // MinValue and MaxValue are more general purpose than just "numbers", they should probably live on some common base type
        // Likewise, the concept is a bad fit for types like BigInteger which aren't fixed sized
        // There may also be concern around the primitive types which already define these names as constants

        static abstract TSelf MinValue { get; }

        static abstract TSelf MaxValue { get; }

        static abstract TSelf One { get; }

        static abstract TSelf Zero { get; }

        // Abs mirrors Math.Abs and returns the same type. This can fail for MinValue of signed integer types
        // Swift has an associated type that can be used here, which would require an additional type parameter in .NET
        // However that would hinder the reusability of these interfaces in constraints

        static abstract TSelf Abs(TSelf value);

>         static TSelf Clamp(TSelf value, TSelf min, TSelf max);
> 
>         static (TSelf Quotient, TSelf Remainder) DivRem(TSelf left, TSelf right);
> 
>         static TSelf Max(TSelf x, TSelf y);
> 
>         static TSelf Min(TSelf x, TSelf y);

Reviewer 3: `abstract` is missing from these members.

```suggestion
        static abstract TSelf Clamp(TSelf value, TSelf min, TSelf max);

        static abstract (TSelf Quotient, TSelf Remainder) DivRem(TSelf left, TSelf right);

        static abstract TSelf Max(TSelf x, TSelf y);

        static abstract TSelf Min(TSelf x, TSelf y);
```

        // TODO: We may want to expose constants related to the subnormal and finite boundaries

        // The following constants are defined by System.Math and System.MathF today

        static abstract TSelf E { get; }

        static abstract TSelf PI { get; }

        static abstract TSelf Tau { get; }

        // The following methods are defined by System.Math and System.MathF today
        // Exposing them on the interfaces means they will become available as statics on the primitive types
        // This, in a way, obsoletes Math/MathF and brings float/double inline with how non-primitive types support similar functionality
        // API review will need to determine if we'll want to continue adding APIs to Math/MathF in the future

        static TSelf Acos(TSelf x);

        static TSelf Acosh(TSelf x);

        static TSelf Asin(TSelf x);

        static TSelf Asinh(TSelf x);

        static TSelf Atan(TSelf x);

        static TSelf Atan2(TSelf y, TSelf x);

        static TSelf Atanh(TSelf x);

        static TSelf BitIncrement(TSelf x);

        static TSelf BitDecrement(TSelf x);

        static TSelf Cbrt(TSelf x);

>         static TSelf Ceiling(TSelf x);
> 
>         // CopySign is likely more general purpose and may apply to any signed number type
>         // It may even be applicable to unsigned numbers where it simply returns x, for convenience
> 
>         static TSelf CopySign(TSelf x, TSelf y);
> 
>         static TSelf Cos(TSelf x);
> 
>         static TSelf Cosh(TSelf x);
> 
>         static TSelf Exp(TSelf x);
> 
>         static TSelf Floor(TSelf x);
> 
>         static TSelf FusedMultiplyAdd(TSelf x, TSelf y, TSelf z);
> 
>         static TSelf IEEERemainder(TSelf x, TSelf y);
> 
>         // IEEE defines the result to be an integral type, but not the size
> 
>         static int ILogB(TSelf x);
> 
>         static TSelf Log(TSelf x);
> 
>         static TSelf Log(TSelf x, TSelf newBase);
> 
>         static TSelf Log2(TSelf x);
> 
>         static TSelf Log10(TSelf x);
> 
>         static TSelf MaxMagnitude(TSelf x, TSelf y);
> 
>         static TSelf MinMagnitude(TSelf x, TSelf y);
> 
>         static TSelf Pow(TSelf x, TSelf y);
> 
>         static TSelf Round(TSelf x);
> 
>         static TSelf Round(TSelf x, int digits);
> 
>         static TSelf Round(TSelf x, MidpointRounding mode);
> 
>         static TSelf Round(TSelf x, int digits, MidpointRounding mode);
> 
>         // IEEE defines n to be an integral type, but not the size
> 
>         static TSelf ScaleB(TSelf x, int n);
> 
>         static TSelf Sin(TSelf x);
> 
>         static TSelf Sinh(TSelf x);
> 
>         static TSelf Sqrt(TSelf x);
> 
>         static TSelf Tan(TSelf x);
> 
>         static TSelf Tanh(TSelf x);
> 
>         static TSelf Truncate(TSelf x);

Reviewer 3: `abstract` is missing from these members.

```suggestion
        static abstract TSelf Acos(TSelf x);

        static abstract TSelf Acosh(TSelf x);

        static abstract TSelf Asin(TSelf x);

        static abstract TSelf Asinh(TSelf x);

        static abstract TSelf Atan(TSelf x);

        static abstract TSelf Atan2(TSelf y, TSelf x);

        static abstract TSelf Atanh(TSelf x);

        static abstract TSelf BitIncrement(TSelf x);

        static abstract TSelf BitDecrement(TSelf x);

        static abstract TSelf Cbrt(TSelf x);

        static abstract TSelf Ceiling(TSelf x);

        // CopySign is likely more general purpose and may apply to any signed number type
        // It may even be applicable to unsigned numbers where it simply returns x, for convenience

        static abstract TSelf CopySign(TSelf x, TSelf y);

        static abstract TSelf Cos(TSelf x);

        static abstract TSelf Cosh(TSelf x);

        static abstract TSelf Exp(TSelf x);

        static abstract TSelf Floor(TSelf x);

        static abstract TSelf FusedMultiplyAdd(TSelf x, TSelf y, TSelf z);

        static abstract TSelf IEEERemainder(TSelf x, TSelf y);

        // IEEE defines the result to be an integral type, but not the size

        static abstract int ILogB(TSelf x);

        static abstract TSelf Log(TSelf x);

        static abstract TSelf Log(TSelf x, TSelf newBase);

        static abstract TSelf Log2(TSelf x);

        static abstract TSelf Log10(TSelf x);

        static abstract TSelf MaxMagnitude(TSelf x, TSelf y);

        static abstract TSelf MinMagnitude(TSelf x, TSelf y);

        static abstract TSelf Pow(TSelf x, TSelf y);

        static abstract TSelf Round(TSelf x);

        static abstract TSelf Round(TSelf x, int digits);

        static abstract TSelf Round(TSelf x, MidpointRounding mode);

        static abstract TSelf Round(TSelf x, int digits, MidpointRounding mode);

        // IEEE defines n to be an integral type, but not the size

        static abstract TSelf ScaleB(TSelf x, int n);

        static abstract TSelf Sin(TSelf x);

        static abstract TSelf Sinh(TSelf x);

        static abstract TSelf Sqrt(TSelf x);

        static abstract TSelf Tan(TSelf x);

        static abstract TSelf Tanh(TSelf x);

        static abstract TSelf Truncate(TSelf x);
```

        // The following members are exposed on the floating-point types as constants today
        // This may be of concern when implementing the interface

        static abstract TSelf Epsilon { get; }

        static abstract TSelf NaN { get; }

        static abstract TSelf NegativeInfinity { get; }

        static abstract TSelf PositiveInfinity { get; }

        // The following methods are exposed on the floating-point types today

>         static bool IsFinite(TSelf value);
> 
>         static bool IsInfinity(TSelf value);
> 
>         static bool IsNaN(TSelf value);
> 
>         static bool IsNegative(TSelf value);
> 
>         static bool IsNegativeInfinity(TSelf value);
> 
>         static bool IsNormal(TSelf value);
> 
>         static bool IsPositiveInfinity(TSelf value);
> 
>         static bool IsSubnormal(TSelf value);

Reviewer 3: ```suggestion
        static abstract bool IsFinite(TSelf value);

        static abstract bool IsInfinity(TSelf value);

        static abstract bool IsNaN(TSelf value);

        static abstract bool IsNegative(TSelf value);

        static abstract bool IsNegativeInfinity(TSelf value);

        static abstract bool IsNormal(TSelf value);

        static abstract bool IsPositiveInfinity(TSelf value);

        static abstract bool IsSubnormal(TSelf value);
```

        static abstract TSelf MaxMagnitudeNumber(TSelf x, TSelf y);

        static abstract TSelf MaxNumber(TSelf x, TSelf y);

        static abstract TSelf MinMagnitudeNumber(TSelf x, TSelf y);

        static abstract TSelf MinNumber(TSelf x, TSelf y);

        static abstract TSelf Root(TSelf x, TSelf n);

        static abstract TSelf SinPi(TSelf x);

        static abstract TSelf TanPi(TSelf x);

        // The majority of the IEEE required operations are listed below
        // This doesn't include the recommended operations such as sin, cos, acos, etc

        // TODO: There are a couple methods below not covered in the list above, we should
        // review these and determine if they should be implemented

        // 5.3.1
        //  TSelf RoundToIntegral(TSelf x);
        //      TiesToEven
        //      TiesToAway
        //      TowardZero
        //      TowardPositive
        //      TowardNegative
        //      Exact
        //  TSelf NextUp(TSelf x);
        //  TSelf NextDown(TSelf x);
        //  TSelf Remainder(TSelf x, TSelf y);

        // 5.3.3
        //  TSelf ScaleB(TSelf x, TLogBFormat n);
        //  TLogBFormat LogB(TSelf x);

        // 5.4.1

>         //  TSeulf Addition(TSelf x, TSelf y);
>         //  TSeulf Subtraction(TSelf x, TSelf y);
>         //  TSeulf Multiplication(TSelf x, TSelf y);
>         //  TSeulf Division(TSelf x, TSelf y);
>         //  TSeulf SquareRoot(TSelf x);
>         //  TSeulf FusedMultiplyAdd(TSelf x, TSelf y, TSelf z);

Reviewer 12: ```suggestion
        //  TSelf Addition(TSelf x, TSelf y);
        //  TSelf Subtraction(TSelf x, TSelf y);
        //  TSelf Multiplication(TSelf x, TSelf y);
        //  TSelf Division(TSelf x, TSelf y);
        //  TSelf SquareRoot(TSelf x);
        //  TSelf FusedMultiplyAdd(TSelf x, TSelf y, TSelf z);
```
Small typos

    }

    public struct Half : IBinaryFloatingPoint<Half> { }

    public struct Int16 : ISignedBinaryInteger<short> { }

    public struct Int32 : ISignedBinaryInteger<int> { }

    public struct Int64 : ISignedBinaryInteger<long> { }

    public struct IntPtr : ISignedBinaryInteger<nint> { /* Size */ }

    public struct SByte : ISignedBinaryInteger<sbyte> { }

    public struct Single : IBinaryFloatingPoint<float> { }

    public struct TimeSpan
        : IAddable<TimeSpan, TimeSpan>,
          IComparableOperators<TimeSpan, TimeSpan>, // implies IEquatableOperators<TimeSpan, TimeSpan>
          IDivisable<TimeSpan, TimeSpan>,
          IDivisable<TimeSpan, double>,
          IMultipliable<TimeSpan, TimeSpan>,
          IMultipliable<TimeSpan, double>,
          INegatable<TimeSpan>
    {
    }

    public struct UInt16 : IUnsignedBinaryInteger<ushort> { }

    public struct UInt32 : IUnsignedBinaryInteger<uint> { }

    public struct UInt64 : IUnsignedBinaryInteger<ulong> { }

    public struct UIntPtr : IUnsignedBinaryInteger<nuint> { }
}
```

### Pending Concepts

There are several types which may benefit from some interface support. These include, but aren't limited to:

> * System.Array
> * System.Enum
> * System.Index
> * System.Range
> * System.Tuple
> * System.ValueTuple

Reviewer 3: ```suggestion
* System.Array
* System.DateOnly
* System.Enum
* System.Index
* System.Range
* System.TimeOnly
* System.Tuple
* System.ValueTuple
```

> ## Designs

Author: That's fair, I can updated the owner.

I'll see about adding the scenarios we care about, I'm not great at doc writing however, so we'll see how it turns out 😄

> ## Designs

Reviewer 13: > I'm not great at doc writing however

I don't know about that. So far, I see a pretty decent write up of a very non-trivial feature. Don't put yourself down :-)

> The overall design has been influenced by the existing .NET Surface Area, the Swift numeric protocols (https://developer.apple.com/documentation/swift/swift_standard_library/numbers_and_basic_values/numeric_protocols), and the Rust op traits (https://doc.rust-lang.org/std/ops/index.html). It is broken up into a few layers describing a heirarchy of interfaces and what types implement them. The interfaces aim to be reusable by a variety of types and so do not provide guarantees around concepts like associativity, commutativity or distributivity. Instead, these concepts may become available to tooling in the future via another mechanism, such as attributes on the respective operator implementations.

Reviewer: ```suggestion
The overall design has been influenced by the existing .NET Surface Area, the Swift numeric protocols (https://developer.apple.com/documentation/swift/swift_standard_library/numbers_and_basic_values/numeric_protocols), and the Rust op traits (https://doc.rust-lang.org/std/ops/index.html). It is broken up into a few layers describing a hierarchy of interfaces and what types implement them. The interfaces aim to be reusable by a variety of types and so do not provide guarantees around concepts like associativity, commutativity or distributivity. Instead, these concepts may become available to tooling in the future via another mechanism, such as attributes on the respective operator implementations.
```

### Base Interfaces

The base interfaces each describe one or more core operations that are used to build up the core interfaces listed later. In many cases, such as for addition, subtraction, multiplication, and division, there is only a single operator per interface. This split is important in order for them to be reusable against downstream types, like `DateTime` where addition does not imply subtraction or `Matrix` where multiplication does not imply division. These interfaces are not expected to be used by many users and instead primarily by developers looking at defining their own abstractions, as such we are not concerned with adding simplifying interfaces like `IAddable<TSelf> : IAddable<TSelf, TSelf>`. Likewise, there is not a concern with exposing `IAddable<TSelf, TOther, TResult>` and instead we will require `TSelf` and `TResult` be the same. This requirement greatly simplifies implementation and consumption of the types.

> ```csharp

Reviewer 13: Could add this?

```markdown
<!--
@startuml
IParseable <|-- ISpanParseable
IEquatable <|-- IEquatableOperators

IComparable <|-- IComparableOperators
IEquatableOperators <|-- IComparableOperators

IAddable          <|-- INumber
IDecrementable    <|-- INumber
IDivisable        <|-- INumber
IIncrementable    <|-- INumber
IMultipliable     <|-- INumber
IRemainder        <|-- INumber
ISubtractable     <|-- INumber

INumber    <|-- ISignedNumber
INegatable <|-- ISignedNumber

INumber <|-- IUnsignedNumber
INumber <|-- IBinaryNumber

IBinaryNumber <|-- ISignedBinaryNumber
ISignedNumber <|-- ISignedBinaryNumber

IBinaryNumber   <|-- IUnsignedBinaryNumber
IUnsignedNumber <|-- IUnsignedBinaryNumber

IBinaryNumber <|-- IBinaryInteger

IBinaryInteger      <|-- ISignedBinaryInteger
ISignedBinaryNumber <|-- ISignedBinaryInteger

IBinaryInteger        <|-- IUnsignedBinaryInteger
IUnsignedBinaryNumber <|-- IUnsignedBinaryInteger

ISignedNumber <|-- IFloatingPoint

IFloatingPoint <|-- IBinaryFloatingPoint
IBinaryNumber  <|-- IBinaryFloatingPoint

IFloatingPoint <|-- IDecimalFloatingPoint

interface IParseable<TSelf>
interface ISpanParseable<TSelf>
interface IEquatableOperators<TSelf, TOther>
interface IComparableOperators<TSelf, TOther>
interface IAddable<TSelf, TOther>
interface ISubtractable<TSelf, TOther>
interface IIncrementable<TSelf>
interface IDecrementable<TSelf>
interface IMultipliable<TSelf, TOther>
interface IDivisable<TSelf, TOther>
interface IRemainder<TSelf, TOther>
interface INegatable<TSelf>
interface INumber<TSelf>
interface ISignedNumber<TSelf>
interface IUnsignedNumber<TSelf>
interface IBinaryNumber<TSelf>
interface ISignedBinaryNumber<TSelf>
interface IUnsignedBinaryNumber<TSelf>
interface IBinaryInteger<TSelf>
interface ISignedBinaryInteger<TSelf>
interface IUnsignedBinaryInteger<TSelf>
interface IFloatingPoint<TSelf>
interface IBinaryFloatingPoint<TSelf>
interface IDecimalFloatingPoint<TSelf>
@enduml
-->

![UML](http://www.plantuml.com/plantuml/svg/VPHHJiCm44J_VOeSWDwWGWKAaZzeAr8Eu2RROCbnWkCW8N5ufAOCNhHilshFs_7QoivcoYxbkG_67boQo9q2LUklrQgozUWYzenzV9zTBcIizoCbbuSq6MCVXdvqgQ2mKM12gK8RUzUs5qVy5dWtzoTAncwfITHJl3WAil_mKrdFPHlLwkSvP3y63mCklr3lV6mfoURrVCh9DLckDjS_lqBjkqZjJ_MEEhmuKrdneht6gLIvUE-ZIvyeB9VCcd7iE8PfTXNlXxFO5LfNRTazRCpK0S4Qxy_x12uqBz2gkNWvb8WNKIl4DtuAWyi-TeV1nsmCNxE7u2H_C07KBRVK-Dw5FwYFcTBPDLGX7cXzh2cSRqiDKI7hmi0lq4rrtEStIWmMGa2Xh-6W41X25UD18DnXIxgEfD2FGHmf3595OH0EGYlBDo08-Dna6UCf0mWItHwGVeYEyV7JpmAb7wPmV2B-xKQdn1a2jg7Opdtu1W00)
```

This will render a UML diagram:

![UML](http://www.plantuml.com/plantuml/svg/VPHHJiCm44J_VOeSWDwWGWKAaZzeAr8Eu2RROCbnWkCW8N5ufAOCNhHilshFs_7QoivcoYxbkG_67boQo9q2LUklrQgozUWYzenzV9zTBcIizoCbbuSq6MCVXdvqgQ2mKM12gK8RUzUs5qVy5dWtzoTAncwfITHJl3WAil_mKrdFPHlLwkSvP3y63mCklr3lV6mfoURrVCh9DLckDjS_lqBjkqZjJ_MEEhmuKrdneht6gLIvUE-ZIvyeB9VCcd7iE8PfTXNlXxFO5LfNRTazRCpK0S4Qxy_x12uqBz2gkNWvb8WNKIl4DtuAWyi-TeV1nsmCNxE7u2H_C07KBRVK-Dw5FwYFcTBPDLGX7cXzh2cSRqiDKI7hmi0lq4rrtEStIWmMGa2Xh-6W41X25UD18DnXIxgEfD2FGHmf3595OH0EGYlBDo08-Dna6UCf0mWItHwGVeYEyV7JpmAb7wPmV2B-xKQdn1a2jg7Opdtu1W00)

Author: This is great, I didn't know it was supported.

Reviewer 13: It's not really; but you'll see once you inspect the snippet closely :-)

Reviewer 2: Tricky tricky

        static abstract bool operator >(TSelf lhs, TOther rhs);

        static abstract bool operator >=(TSelf lhs, TOther rhs);
    }

    public interface IAddable<TSelf, TOther>
        where TSelf : IAddable<TSelf, TOther>
    {
        // An additive identity is not exposed here as it may be problematic for cases like
        // DateTime that implements both IAddable<DateTime, DateTime> and IAddable<DateTime, TimeSpan>
        // This would require overloading by return type and may be problematic for some scenarios where TOther might not have an additive identity

        // We can't assume TOther + TSelf is valid as not everything is commutative

        // We can't expose TSelf - TOther as some types only support one of the operations

        static abstract TSelf operator +(TSelf value);

        static abstract TSelf operator +(TSelf lhs, TOther rhs);

        // There is an open question on if we should support "checked" versions of the operators
        // Many algorithms, such as in LINQ, depend on checked behavior for overflow
        // Supporting this would require well-defined ECMA-335 names and language work to ensure the right things happen
    }

>     public interface ISubtractable<TSelf, TOther>

Reviewer 14: A little bit surprised by the types here.  I would have expected

```csharp
public interface ISubtractable<TSelf, TDiff>
  where TSelf : ISubtractable<TSelf, TDiff>
{
    static abstract TDiff operator- (TSelf lhs, TSelf rhs);
}
```

that is, a generalization of `DateTime dt1, dt2; TimeSpan diff = dt1 - dt2`, rather than of `DateTime dt1; TimeSpan diff;  DateTime dt2 = dt1 - diff` since the latter can be expressed in terms of addition and negation.

Author: The current design has been that `typeof(TSelf) == typeof(TResult)` because that greatly simplifies things in terms of consumption and matches what the vast majority of types end up following.

There are some interesting edge cases where `TResult` can differ, but that's a minority compared to where `TOther` can differ, and so we have an option of `TSelf, TOther` for simplicity or `TSelf, TOther, TResult` for completeness.
It may be that we do need `TSelf, TOther, TResult` in practice, especially as community feedback comes in and outreach to major libraries happens.

Reviewer 13: I assume this means that people couldn't call do `dateTime1 - dateTime2` from generic code at all, right? Or is there a viable, but slightly clunky way to do that?

Author: Correct, we would need to expose `ISubtractable<TSelf, TOther, TResult>` where `DateTime : ISubtractable<DateTime, DateTime, TimeSpan>`

There may be enough scenarios that its worth doing this, even if its relatively uncommon.

Author: I've changed this to have `ISubtractable<TSelf, TOther, TResult>` as there are enough use cases in System that I think it justifies doing it from day 1, and it ensures we always meet the needs, since its the exact subtraction signature anyways.

        where TSelf : ISubtractable<TSelf, TOther>
    {

>         static abstract TSelf operator -(TSelf lhs, TOther rhs);

Reviewer 11: I like the idea of creating one interface per operator (or group of operators in the case of equality or comparison). It emphasizes the syntactic nature of the protocols versus building maintaining a hierarchy of algebraic structures (and any implied laws).

        // MinValue and MaxValue are more general purpose than just "numbers", they should probably live on some common base type
        // Likewise, the concept is a bad fit for types like BigInteger which aren't fixed sized
        // There may also be concern around the primitive types which already define these names as constants

        static abstract TSelf MinValue { get; }

        static abstract TSelf MaxValue { get; }

        static abstract TSelf One { get; }

        static abstract TSelf Zero { get; }

        // Abs mirrors Math.Abs and returns the same type. This can fail for MinValue of signed integer types
        // Swift has an associated type that can be used here, which would require an additional type parameter in .NET
        // However that would hinder the reusability of these interfaces in constraints

        static abstract TSelf Abs(TSelf value);

        static abstract TSelf Clamp(TSelf value, TSelf min, TSelf max);

        static abstract (TSelf Quotient, TSelf Remainder) DivRem(TSelf left, TSelf right);

        static abstract TSelf Max(TSelf x, TSelf y);

>         static abstract TSelf Min(TSelf x, TSelf y);

Reviewer 11: These methods should be part of `IComparableOperators`.

As discussed, complex numbers or matrix algebras cannot define total orders and as such `Min` and `Max` operators cannot be reasonably defined.

Author: Moving them to `IComparable` wouldn't help with `Complex` or `Matrix` like types, given that `INumber` implements `IComparable`.

It would, however, impact types like `Guid`, `DateTime`, `TimeSpan` and others that don't define `Min`, `Max`, or `Clamp` but would implement `IComparable`. I don't see an issue here right now, but thought I'd call it out.

---

On a meta point, I think the interesting split between `Complex`, `Vector`, `Matrix`, and `INumber` is that the latter really defines something that can be interpreted as a scalar value. A `Vector` is composed of two or more scalars and a `Matrix` is composed of two or more vectors. In that sense, a `Complex` number is most like a vector in that it has two scalar parts (a real part `a` and an imaginary part `bi`).

How `Complex` composes here needs some more thought and prototyping to determine how it can be exposed, particularly given I expect most users will largely care about this from the "scalar" (single value) concept for `INumber` in typical use-cases.

Likewise, while `Matrix` and `Vector` are often not `IComparable` in the traditional sense, they do often support these operations returning `TSelf` for an "element-wise" comparison, which is often used as a mask for downstream operations and sometimes as an "do all elements follow this comparison" (as is the case for `Vector<T>`).

Reviewer 11: > Moving them to IComparable wouldn't help with Complex or Matrix like types, given that INumber implements IComparable.

Right, I was suggesting that perhaps `INumber` should not inherit the comparison interface. 

> It would, however, impact types like Guid, DateTime, TimeSpan and others that don't define Min, Max, or Clamp but would implement IComparable. I don't see an issue here right now, but thought I'd call it out.

It is possible to express `Min`, `Max` and `Clamp` in terms of comparison, so perhaps we could add those as DIMs in the comparison interface.

> On a meta point, I think the interesting split between Complex, Vector, Matrix, and INumber is that the latter really defines something that can be interpreted as a scalar value. A Vector is composed of two or more scalars and a Matrix is composed of two or more vectors. In that sense, a Complex number is most like a vector in that it has two scalar parts (a real part a and an imaginary part bi).

From a memory representation perspective that's a pretty accurate distinction, in fact it's fairly common to define complex numbers using [matrices](https://en.wikipedia.org/wiki/Complex_number#Matrix_representation_of_complex_numbers). From an abstraction perspective however complex numbers are numbers, and an abstraction called `INumber` should reasonably support those as well.

> Likewise, while Matrix and Vector are often not IComparable in the traditional sense, they do often support these operations returning TSelf for an "element-wise" comparison, which is often used as a mask for downstream operations and sometimes as an "do all elements follow this comparison" (as is the case for Vector<T>).

I suspect though that such a comparison would not be a total order and therefore cannot be used meaningfully for things like building a search tree.

Author: > Right, I was suggesting that perhaps INumber should not inherit the comparison interface.

That will hinder what I've currently tracked as the "most common scenario", which is abstracting over the 13 primitive types (`sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `nint`, `nuint`, `long`, `ulong` as well as `float`, `double`, and `half`). The "typical" constraint would then become `where T : INumber<T>, IComparableOperators<T, T>`. Types like `char` and `decimal` are also sometimes included here, as well as types like `BigInteger` and a conceptual future `BigFloat` or types like `Int128`, `UInt128`, `decimal32`, `decimal64`, `decimal128` (the IEEE decimal types) and `Quad` (the IEEE `binary128`). 

Types like `Complex`, while conceptually "numbers", really have a different "interface" and handling when thinking from an OOP programming algorithm. I would think the same applies to things like `Vector` or `Matrix` where even though they support many of the operations, you want to consider and handle them differently and would likely have a different overload to do so.

We could have something like `INumber` vs `IScalarNumber`, but there is cost (including runtime startup cost) to every one of the interfaces here so we need to be careful about making this "too expansive" while still allowing the correct user scenarios.

Reviewer 11: > We could have something like INumber vs IScalarNumber, but there is cost (including runtime startup cost) to every one of the interfaces here so we need to be careful about making this "too expansive" while still allowing the correct user scenarios.

We could perhaps strike a balance by doing away some of the other intermediate interfaces (e.g. `IAddable`). Really though I find this to be a naming issue, we could consider renaming `INumber` to something like `IScalarNumber` or `IPrimitiveNumber` to avoid confusion.

Author: > Really though I find this to be a naming issue

Right, this is roughly based on what Swift calls it (`Numeric` protocol) where they likewise imply similar requirements/guarantees around Comparisons, Equality, etc.
We could clarify by calling it something else or by having an intermediate step between "numbers" and "scalar numbers".

> We could perhaps strike a balance by doing away some of the other intermediate interfaces (e.g. IAddable).

I think that will be too limiting and I'd rather have a strong base here that allows other libraries to define their own contracts as appropriate. Otherwise, we basically end up with `IComparableOperators`, `IEquatableOperators`, and `INumber` (plus potentially something like `IVector`) and so DateTime and others won't really be supported here.

Reviewer 11: Do we have a use case for DateTime and friends benefiting from generic arithmetic?

Author: Not strictly speaking. They are interesting as they are types where a user may want to operate over the lower base interfaces (e.g. IAddable) rather than the higher abstractions (e.g. INumber).

You could imagine a `Sum<T> where T : IAddable<T, T>` method or similar being a valid use-case here.

Author: The same applies for the general comparison, equality, min, max, and similar functionality that could be defined for types like DateTime, TimeSpan, Guid, etc.

Reviewer 11: > You could imagine a Sum<T> where T : IAddable<T, T> method or similar being a valid use-case here.

Right, so effectively defining reduce over a monoid-like abstraction. In practice I find that the definition of addition greatly varies depending on the application, and it might often not coincide with what the type actually implements (e.g. reduce using multiplication or some other binary operation). It is usually simpler to just use LINQ's `Aggregate`.

> The same applies for the general comparison, equality, min, max, and similar functionality that could be defined for types like DateTime, TimeSpan, Guid, etc.

There's certainly value in doing that. I'm just not convinced that addition is as useful for those particular types.

Author: > It is usually simpler to just use LINQ's Aggregate.

Many of these concepts came up as a question of "if we were to implement LINQ today, would they take `T where T : IAddable<T>`"

I imagine most algorithms don't care that it is a monoid, they just care that you can do:
```csharp
T sum = default; 

foreach (var value in values)
{
    sum += value;
}

return sum;
```

This like commutativity, associativity, distributivity and others (particularly in programming) largely only come up in the context of optimizations (like constant folding). They almost never come up otherwise (outside maybe simplification, which would be an analyzer, not something C# does itself).

Reviewer 11: I wasn't really making a point about monoid laws, and you're right that most simple reduce operations don't depend on them holding. I'm merely questioning the usefulness of `T where T : IAddable<T>` methods in the grander scheme of things.

Author: It's definitely possible that in practice they won't be used and that's something we'll need to determine before shipping.

At this time, and based on other languages/frameworks, they seem like good/practical building blocks that we'll want to expose.

Will definitely keep this in mind as the community outreach begins.

    }

    public struct Decimal
        : ISignedNumber<decimal>
    {
        // Decimal defines a few additional operations like Ceiling, Floor, Round, and Truncate
        // The rest of the IEEE operations are missing.

        // Decimal and some other types likewise define "friendly" names for operators, should we expose them?
    }

    public struct Double : IBinaryFloatingPoint<double> { }

    public struct Guid
        : IComparableOperators<Guid, Guid> // implies IEquatableOperators<Guid, Guid>
    {
    }

    public struct Half : IBinaryFloatingPoint<Half> { }

    public struct Int16 : ISignedBinaryInteger<short> { }

    public struct Int32 : ISignedBinaryInteger<int> { }

    public struct Int64 : ISignedBinaryInteger<long> { }

    public struct IntPtr : ISignedBinaryInteger<nint> { /* Size */ }

    public struct SByte : ISignedBinaryInteger<sbyte> { }

    public struct Single : IBinaryFloatingPoint<float> { }

    public struct TimeSpan
        : IAddable<TimeSpan, TimeSpan>,
          IComparableOperators<TimeSpan, TimeSpan>, // implies IEquatableOperators<TimeSpan, TimeSpan>

>           IDivisable<TimeSpan, TimeSpan>,

Reviewer 15: Doesn't this return a double, not a TimeSpan?

Author: Yes, I misread the docs.

Such an API would either require `IDivisable<TSelf, TOther, TResult>` or not being exposed via one of these interfaces. Exposing a `TResult` has been avoided so far to simplify things and make the APIs more usable in the most common case.

Author: I've changed this to have `IDivisable<TSelf, TOther, TResult>` as there are enough use cases in System that I think it justifies doing it from day 1

          IDivisable<TimeSpan, double>,
          IMultipliable<TimeSpan, TimeSpan>,
          IMultipliable<TimeSpan, double>,
          INegatable<TimeSpan>
    {
    }

    public struct UInt16 : IUnsignedBinaryInteger<ushort> { }

    public struct UInt32 : IUnsignedBinaryInteger<uint> { }

    public struct UInt64 : IUnsignedBinaryInteger<ulong> { }

    public struct UIntPtr : IUnsignedBinaryInteger<nuint> { }
}
```

### Pending Concepts

There are several types which may benefit from some interface support. These include, but aren't limited to:
* System.Array
* System.DateOnly
* System.Enum
* System.Index
* System.Range
* System.TimeOnly
* System.Tuple

> * System.ValueTuple

Reviewer 8: ValueTuple is instantiated a lot. Adding more interfaces to it is very problematic performance-wise.

* System.Numerics.BigInteger
* System.Numerics.Complex
* System.Numerics.Matrix3x2
* System.Numerics.Matrix4x4
* System.Numerics.Plane
* System.Numerics.Quaternion
* System.Numerics.Vector<T>
* System.Numerics.Vector2
* System.Numerics.Vector3
* System.Numerics.Vector4
* System.Runtime.Intrinsics.Vector64<T>
* System.Runtime.Intrinsics.Vector128<T>
* System.Runtime.Intrinsics.Vector256<T>

Likewise, there are many comments within the defined interfaces above that call out key points that need additional modeling/consideration. Several of these concepts will be determined via user research and outreach while others will be determined via API review and feedback from other area experts. Others may be determined by language or runtime limitations in what is feasible for them to support.

## Requirements

> ## Scenarios

Author: I pushed a couple scenarios, if you could review and indicate if they roughly follow what you expect, that'd be much appreciated 😄
